lib/tapioca/compilers/dsl/active_record_columns.rb

```diff
@@ -0,0 +1,347 @@
+          )
+          add_method(
+            klass,
+            "#{column_name}_changed?",
+            methods_to_add,
+            return_type: "T::Boolean"
+          )
+          add_method(
+            klass,
+            "#{column_name}_will_change!",
+            methods_to_add
+          )
+          add_method(
+            klass,
+            "#{column_name}_was",
+            methods_to_add,
+            return_type: getter_type
+          )
+          add_method(
+            klass,
+            "#{column_name}_previous_change",
+            methods_to_add,
+            return_type: "[#{getter_type}, #{getter_type}]"
+          )
+          add_method(
+            klass,
+            "#{column_name}_previously_changed?",
+            methods_to_add,
+            return_type: "T::Boolean"
+          )
+          add_method(
+            klass,
+            "#{column_name}_previously_was",
+            methods_to_add,
+            return_type: getter_type
+          )
+          add_method(
+            klass,
+            "restore_#{column_name}!",
+            methods_to_add
+          )
+
+          # Added by ActiveRecord::AttributeMethods::BeforeTypeCast
+          #
+          add_method(
+            klass,
+            "#{column_name}_before_type_cast",
+            methods_to_add,
+            return_type: "T.untyped"
+          )
+          add_method(
+            klass,
+            "#{column_name}_came_from_user?",
+            methods_to_add,
+            return_type: "T::Boolean"
+          )
+        end
+
+        sig do
+          params(
+            constant: T.class_of(ActiveRecord::Base),
+            column_name: String
+          ).returns([String, String])
+        end
+        def type_for(constant, column_name)
+          column_type = constant.attribute_types[column_name]
+
+          if Object.const_defined?(:StrongTypeGeneration) &&
+              !(constant.singleton_class < Object.const_get(:StrongTypeGeneration))
+            return ["T.untyped", "T.untyped"]
+          end
+
+          getter_type =
+            case column_type
+            when ActiveRecord::Type::Integer
+              "::Integer"
+            when ActiveRecord::Type::String
+              "::String"
+            when ActiveRecord::Type::Date
+              "::Date"
+            when ActiveRecord::Type::Decimal
+              "::BigDecimal"
+            when ActiveRecord::Type::Float
+              "::Float"
+            when ActiveRecord::Type::Boolean
+              "T::Boolean"
+            when ActiveRecord::Type::DateTime, ActiveRecord::Type::Time
+              "::DateTime"
+            when ActiveRecord::AttributeMethods::TimeZoneConversion::TimeZoneConverter
+              "::ActiveSupport::TimeWithZone"
+            else
+              "T.untyped"
+            end
+
+          column = constant.columns_hash[column_name]
+          setter_type = getter_type
+
+          if column&.null ||
```

Comment:
Not sure if we are testing these conditions, but it would be nice to ensure that a field that is marked as `null: true` and `null: false` creates different return types and that `created_at`/`updated_at` columns and primary key columns are always nilable.

Reply:
It maybe how I'm running the tests but we don't actually generate any methods for `created_at/updated_at`

Reply:
I don't see us generating those columns, though.

Reply:
`t.timestamps`

Reply:
I thought `created_at/updated_at` came automatically with every AR object regardless of the schema. I guess it's more so that `t.timestamps` is inserted automatically when an object is created.

Reply:
Hey @KaanOzkan I was wondering why primary_keys are set as nilable. Other columns with null: false are not nilable, so I was expecting the same for primary_key columns

Reply:
@alse The situation is more complicated, and, IMO, we've made a mistake with how we generate other column types, not primary keys.

In general an Active Record model would have all of its fields nilable:
```ruby
class User < ActiveRecord::Base
  # user has "id" and "name" fields
end

user = User.new
# => #<User:0x00007fefed98b090 id: nil, name: nil>
```

This is because AR models actually have a dual existence. Any model that is not persisted has no guarantees to the types of any of its fields, they can all be nil. This is most acute for the primary key field value which is almost never supplied as a parameter to the `.new` call. Only persisted models actually get the benefit of all the validations and the not-null annotations etc.

For a similar reason, we'd changed all AR associations to be nilable, we might actually do the same for columns as well, since that is actually the more correct way to type an AR model's column attributes under all conditions.

Reply:
That makes sense, thanks for thanks for the background @paracycle !